### PR TITLE
Add argument count mismatch diagnostic

### PR DESCRIPTION
**Pull request by jonas-schievink**
Closes https://github.com/rust-analyzer/rust-analyzer/issues/4025.

This currently has one false positive on this line, where `max` is resolved to `Iterator::max` instead of `Ord::max`:

https://github.com/rust-analyzer/rust-analyzer/blob/8aa10c00a4c5b957d459fac5a103cd9688e8dcdd/crates/expect/src/lib.rs#L263

(I have no idea why it thinks that `usize` is an `Iterator`)

TODO:
* [x] Tests
* [x] Improve diagnostic text for method calls

### CODE REVIEW

**Comment by matklad** [comment]
>This currently has one false positive on this line, where max is resolved to Iterator::max instead of Ord::max:

goto def goes to `Ord` though

**Comment by flodiebold** [comment]
> (I have no idea why it thinks that usize is an Iterator)

Without having looked into it, I'll guess:
 - at that point it's an integer variable
 - resolving the `Ord` impl may require some of the missing integer variable handling
 - for the method to be resolved, we don't need to positively prove that the type implements `Iterator`, but just be unable to prove that it doesn't. With the many impls that exist for `Iterator`, that can often happen.

**Comment by flodiebold** [comment]
As for the PR, I wonder whether the diagnostic should just be produced during inference, but I think it's fine to have it separate for now as well.

**Comment by jonas-schievink** [comment]
@flodiebold Ah, okay, that's surprising. I thought that methods only get resolved in r-a when the obligation is actually proven. Kind of makes sense for an IDE though.

**Comment by flodiebold** [comment]
@jonas-schievink it's how rustc works as well, as far as I understand -- method selection just checks whether the obligation [*may* hold](https://github.com/rust-lang/rust/blob/1d919c9377f4602d991ca1c7ba852e7555943740/src/librustc_typeck/check/method/probe.rs#L1383) (i.e. it's at least ambiguous). At the time of method selection, there's often not enough information to conclusively decide whether the obligation holds.

**Comment by jonas-schievink** [comment]
Hmm, yeah, that's true. rustc will prove the obligation and report an error if it doesn't hold though. Is there any way I can tell that this happened in r-a?

**Comment by flodiebold** [comment]
We currently don't keep track of which obligations failed (that would happen [here](https://github.com/rust-analyzer/rust-analyzer/blob/7566d7da8ab035a286f82a60b8e2c679158bc56c/crates/ra_hir_ty/src/infer.rs#L343)), and even if we did, we'd also need to keep track of where the obligation came from to connect this back to the method.

I wonder whether the false positive is acceptable here in a sense though, because this situation also means that goto def will go to the wrong function, and the return value will potentially be wrong, so we're already noticeably wrong anyway... Or if we're worried about false positives, we could just do the diagnostic for non-trait methods and function calls, which should be less likely to be wrong. But I am in favour of starting to have more diagnostics based on inference. I think it'd be great to have an option for more experimental, more-likely-to-be-false-positive diagnostics.

File: crates/ra_hir_ty/src/expr.rs [hunk 1]
```diff
@@ -312,3 +386,132 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:391]: I suggest using `expect` crate for new test modules (and for new tests in general) :-)

Comment (matklad) [reply crates/ra_hir_ty/src/expr.rs:391]: Specifically, the pattern that I find works best is defining and

```rust
fn check(ra_fixture: &str, expect: Expt) {
}

function in the test module, and then every test just calls this one. 
```

File: crates/ra_hir_ty/src/expr.rs [hunk 2]
```diff
@@ -312,3 +386,132 @@ pub fn record_pattern_missing_fields(
...
+
+    #[test]
+    fn simple_free_fn_zero() {
+        assert_snapshot!(check_diagnostic_message(
+            r"
+            fn zero() {}
+
+            fn f() {
+                zero(1);
+            }
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:416]: I suggest using a more "compressed" style for tests: 

```rust
fn zero() {}
fn main() { zero(1); }
```

Seems like for small snippets extra newlines don't increase visibility, but they increase the overall percieved verbosity of tests

File: crates/ra_hir_ty/src/expr.rs [hunk 3]
```diff
@@ -138,6 +148,70 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         }
     }
 
+    fn validate_call(&mut self, db: &dyn HirDatabase, call_id: ExprId, expr: &Expr) -> Option<()> {
+        // Check that the number of arguments matches the number of parameters.
+
+        // Due to shortcomings in the current type system implementation, only emit this diagnostic
+        // if there are no type mismatches in the containing function.
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:155]: :+1:

File: crates/ra_hir_ty/src/expr.rs [hunk 4]
```diff
@@ -138,6 +148,70 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
...
+        if self.infer.type_mismatches.iter().next().is_some() {
+            return Some(());
+        }
+
+        let is_method_call;
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:160]:
```suggestion
        let is_method_call = matches!(expr, Expr::MethodCall { .. });
```

File: crates/ra_hir_ty/src/expr.rs [hunk 5]
```diff
@@ -138,6 +148,70 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
...
+        };
+
+        let loc = callee.lookup(db.upcast());
+        let ast = loc.source(db.upcast());
+        let params = ast.value.param_list()?;
+
+        let mut param_count = params.params().count();
+        let mut arg_count = args.len();
+
+        if params.self_param().is_some() {
```
Comment (matklad) [inline crates/ra_hir_ty/src/expr.rs:192]: Urgh, we need to already invent some idiom for two param counts, there are many instances of `+1` code

**Pull request merged**